python/pyarrow/dataset.py

```diff
@@ -798,6 +800,20 @@ def write_dataset(data, base_dir, basename_template=None, format=None,
         used determined by the number of available CPU cores.
     max_partitions : int, default 1024
         Maximum number of partitions any batch may be written into.
+    max_open_files : int, default 1024
+        Maximum number of number of files can be opened
```

Comment:
I found this a little confusing. Could we add the full docstring from the C++ docs?

```suggestion
        Maximum number of number of files that can be open at a time. 
        If an attempt is made to open too many files then the least recently 
        used file will be closed. If this setting is set too low you may end up
        fragmenting your data into many small files.
```

Reply:
Yes, I understand. Added the C++ full docstrings. 

> Went through the tests. I had a few questions on some parts that I didn't understand, but I think they generally seem right.
> 
> If you are up for the task, I think it may be worth consolidating them so it's simpler to read through.

Of course, I will improve the test cases. Thank you for the detailed review comments.